src/foam/dao/Relationship.js

```diff
@@ -589,76 +628,68 @@ foam.CLASS({
       name: 'targetDAOKey'
     },
     'targetProperty',
-    {
-      name: 'expression',
-      factory: function() {
-        var sourceProperty = this.sourceProperty;
-        var targetProperty = this.targetProperty;
-        var targetDAOKey   = this.targetDAOKey;
-        var junctionDAOKey = this.junctionDAOKey;
-        var junction       = this.junction;
+  ],
 
-        return function(id) {
+  methods: [
+    function installInClass(cls) {
+      var sourceProperty = this.sourceProperty;
+      var targetProperty = this.targetProperty;
+      var targetDAOKey   = this.targetDAOKey;
+      var junctionDAOKey = this.junctionDAOKey;
+      var junction       = this.junction;
+      var methodName     = this.methodName;
+
+      cls.installAxiom(foam.core.Method.create({
+        name: methodName,
+        args: [
+          {
+            name: 'x',
+            javaType: 'foam.core.X',
+            swiftType: 'Context',
+          }
+        ],
+        returns: 'foam.dao.ManyToManyRelationship',
+        code: function(x) {
           return foam.dao.ManyToManyRelationshipImpl.create({
-            sourceId: id,
+            sourceId: this.id,
             sourceProperty: sourceProperty,
             targetProperty: targetProperty,
             targetDAOKey: targetDAOKey,
             junctionDAOKey: junctionDAOKey,
             junction: junction
-          }, this);
-        };
-      }
-    }
-  ],
-
-  methods: [
-    function buildJavaClass(cls) {
-      this.SUPER(cls);
-      cls.method({
-        name: this.name,
-        visibility: 'public',
-        type: 'foam.dao.ManyToManyRelationship',
-        args: [{ name: 'x', type: 'foam.core.X' }],
-        body: `return new foam.dao.ManyToManyRelationshipImpl.Builder(x)
```

Comment:
Could we just add this as javaCode to the method we add in JS?

I guess then we would be generating this code when running in web mode so maybe that's not the best idea.

Reply:
Nevermind my idea appears to be what you did, I just didn't read the diff very well.  github needs a colourblind mode because red/green diffs are not fun when your display brightness isn't stellar.